### PR TITLE
Transaction Queue Integration

### PR DESCRIPTION
**Pull request by tomusdrw**

### CODE REVIEW

**Comment by arkpar** [comment]
See comments from https://github.com/ethcore/parity/pull/609

File: sync/src/chain.rs [hunk 1]
```diff
@@ -1248,6 +1263,37 @@ impl ChainSync {
...
+			let mut transaction_queue = self.transaction_queue.lock().unwrap();
+			let hashes = txs.iter().map(|tx| tx.hash()).collect::<Vec<H256>>();
+			transaction_queue.remove_all(&hashes, |a| chain.nonce(a));
+		});
+		retracted.for_each(|txs| {
+			// populate sender
+			for tx in &txs {
+				let _sender = tx.sender();
+			}
+			let mut transaction_queue = self.transaction_queue.lock().unwrap();
```
Comment (tomusdrw) [inline sync/src/chain.rs]: @arkpar This happens in multiple threads using `rayon` - note `par_iter`.

File: sync/src/chain.rs [hunk 2]
```diff
@@ -209,6 +212,8 @@ pub struct ChainSync {
 	max_download_ahead_blocks: usize,
 	/// Network ID
 	network_id: U256,
+	/// Transactions Queue
+	transaction_queue: Mutex<TransactionQueue>,
```
Comment (tomusdrw) [inline sync/src/chain.rs]: Transaction queue is not multithreaded. Methods `import` and `remove` internally would have to lock everythng anyway. I'm not sure if we would gain anything from making it thread safe.

What are the most expensive operations? Do you think that instead of removing / adding right away it should just maintain a thread pool internaly?

File: sync/src/chain.rs [hunk 3]
```diff
@@ -921,8 +928,18 @@ impl ChainSync {
...
 	/// Called when peer sends us new transactions
-	fn on_peer_transactions(&mut self, _io: &mut SyncIo, _peer_id: PeerId, _r: &UntrustedRlp) -> Result<(), PacketDecodeError> {
-		Ok(())
+	fn on_peer_transactions(&mut self, io: &mut SyncIo, peer_id: PeerId, r: &UntrustedRlp) -> Result<(), PacketDecodeError> {
+		let chain = io.chain();
+		let item_count = r.item_count();
+		trace!(target: "sync", "{} -> Transactions ({} entries)", peer_id, item_count);
+		let fetch_latest_nonce = |a : &Address| chain.nonce(a);
+
+		let mut transaction_queue = self.transaction_queue.lock().unwrap();
```
Comment (tomusdrw) [inline sync/src/chain.rs]: @arkpar Fixed

File: sync/src/chain.rs (hunk 2 above)
Comment (arkpar) [inline sync/src/chain.rs]: I was asking about thread safety, not multithreaded processing. Already checked that it is not currently thread safe, so mutex is fine here. As for parallelization, I don't currently see the need for that with the transaction volume we have on the main network. Can be added later.

**Comment by gavofyork** [comment]
code quality good. tests good. haven't yet done deep review of logic. might be worth doing a little additional documentation?

**Comment by tomusdrw** [comment]
What kind of documentation would you like? Doctests or just module level description of transaction queue? or specific part of integration?

File: sync/src/lib.rs [hunk 4]
```diff
@@ -153,8 +153,14 @@ impl NetworkProtocolHandler<SyncMessage> for EthSync {
 	}
 
 	fn message(&self, io: &NetworkContext<SyncMessage>, message: &SyncMessage) {
-		if let SyncMessage::BlockVerified = *message {
-			self.sync.write().unwrap().chain_blocks_verified(&mut NetSyncIo::new(io, self.chain.deref()));
+		match *message {
+			SyncMessage::BlockVerified => {
```
Comment (arkpar) [inline sync/src/lib.rs]: `sync` should not care about `BlockVerified` event at all. Please merge `chain_blocks_verified` function into `chain_new_blocks`. These two should be called for the same event and `chain_blocks_verified` is an incorrect name.

Comment (tomusdrw) [inline sync/src/lib.rs]: done

**Comment by arkpar** [comment]
looks gootd to me

File: ethcore/src/client.rs [hunk 5]
```diff
@@ -391,10 +390,16 @@ impl<V> Client<V> where V: Verifier {
 			};
 
 			// Commit results
+			let closed_block = closed_block.unwrap();
+			let receipts = closed_block.block().receipts().clone();
 			closed_block.drain()
 				.commit(header.number(), &header.hash(), ancient)
 				.expect("State DB commit failed.");
 
+			// And update the chain
```
Comment (gavofyork) [inline ethcore/src/client.rs]: why is this moved down here?

File: ethcore/src/service.rs [hunk 6]
```diff
@@ -31,6 +31,8 @@ pub enum SyncMessage {
 		good: Vec<H256>,
 		/// Hashes of blocks not imported to blockchain
 		bad: Vec<H256>,
+		/// Hashes of blocks that were removed from canonical chain
```
Comment (gavofyork) [inline ethcore/src/service.rs]: what if it is retracted and then later on becomes canonical.

File: sync/src/tests/helpers.rs [hunk 7]
```diff
@@ -420,8 +455,8 @@ impl TestNet {
 		self.peers.iter().all(|p| p.queue.is_empty())
 	}
 
-	pub fn trigger_block_verified(&mut self, peer_id: usize) {
+	pub fn trigger_chain_new_blocks(&mut self, peer_id: usize) {
```
Comment (gavofyork) [inline sync/src/tests/helpers.rs]: new blocks in canonical chain? would this be better named `chain_head_changed`?

**Comment by gavofyork** [comment]
several questions. looks good otherwise.

File: ethcore/src/client.rs (hunk 5 above)
Comment (tomusdrw) [inline ethcore/src/client.rs]: It was introduced here: https://github.com/ethcore/parity/commit/bcaed67eaa0257c8998925287fd2dbc40df12698
to avoid race conditions when chain contains uncommitted hashes. Will add a comment explaining the order.

File: ethcore/src/service.rs (hunk 6 above)
Comment (tomusdrw) [inline ethcore/src/service.rs]: in #645 `insert_block` returns retracted blocks and blocks that became canon. Will add more fields here when both are merged.

File: sync/src/tests/helpers.rs (hunk 7 above)
Comment (tomusdrw) [inline sync/src/tests/helpers.rs]: If I understand correctly it doesn't necessarily mean that blocks went to canon chain. I'm open to any different name, though.
